DownwardMetrics: Adds virtio-serial port method (#719)
kubevirt/user-guide: docs/virtual_machines/disks_and_volumes.md

example is SAP HANA.

In order to expose the `downwardMetrics` to the VMs, there are two methods supported: `disk` and `virtio-serial port`.

> **Note:** The **DownwardMetrics** feature gate
> [must be enabled](../operations/activating_feature_gates.md#how-to-activate-a-feature-gate)
> to use the metrics. Available starting with KubeVirt v0.42.0.
 
#### Disk
A volume is created, and it is exposed to the guest as a raw block volume.

Example:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-fedora
  name: vmi-fedora
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: metrics
    machine:
      type: ""
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - name: containerdisk
    containerDisk:
      image: quay.io/containerdisks/fedora:latest
  - name: metrics
    downwardMetrics: {}
```

#### Virtio-serial port
It creates a virtio-serial port named `org.github.vhostmd.1` inside the VM. The metrics may be requested and fetched 

> through this port.

Reviewer: I prefer to invert the order of this, something like (not exact words) "This method uses a virtio-serial port to expose the metrics to the VM..." and then explain the name of the port. Also, I think is good idea to mention that the vhostmd documentation call this "Virtio transport"

Author: Done.

> To enable the `downwardMetrics` within a virtio-serial port, the spec `spec.domain.devices.downwardMetrics: {}` has to be

Reviewer: Maybe it's good idea to say that you need to add a  `downwardMetrics`  _device_, and then the spec as clarification

Author: Agree. Done.

added to the VM definition.

Example:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-fedora
  name: vmi-fedora
spec:
  domain:
    devices:
      downwardMetrics: {}
      disks:
      - disk:
          bus: virtio
        name: containerdisk
    machine:
      type: ""
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - name: containerdisk
    containerDisk:
      image: quay.io/containerdisks/fedora:latest
```

> The `vm-dump-metrics` tool can be used to read the `downwardMetrics` shared with both methods described:

Reviewer: I suggest making this its own section, like `### Accessing metrics data` (or `Reading metrics data`) or something like that, instead of being inside `#### Virtio-serial port`. 

Also, it will be helpful to note that usually the application will connect by itself to the virtio-serial port, and the `vm-dump-metrics` tool is just a standalone tool to access this metrics

Author: Done.

```xml
$ sudo dnf install -y vm-dump-metrics

> $ sudo vm-dump-metrics

Reviewer: Sorry, I don't remember, are root permissions required to access?

Author: Yes, otherwise:
```sh
$ vm-dump-metrics --virtio
LIBMETRICS: get_virtio_metrics(): Unable to export metrics: open(/dev/virtio-ports/org.github.vhostmd.1) Permission denied
LIBMETRICS: get_virtio_metrics(): Unable to read metrics
```

Reviewer: Ok thanks

<metrics>
  <metric type="string" context="host">
    <name>HostName</name>
    <value>node01</value>
[...]
  <metric type="int64" context="host" unit="s">
    <name>Time</name>
    <value>1619008605</value>
  </metric>
  <metric type="string" context="host">
    <name>VirtualizationVendor</name>
    <value>kubevirt.io</value>
  </metric>
</metrics>
```

> **Note:** The tool `vm-dump-metrics` provides the option `--virtio` in case the virtio-serial port is used. 

> > Within the full command `vm-dump-metrics --virtio`, the `DownwardMetrics` can be fetched directly from virtio-serial 

Reviewer: This phrase is not clear. You can also refer to  `vm-dump-metrics --help` for help.

Author: I've dropped some details of the sentence and included the reference to the help info.

### downwardMetrics

> The `downwardMetrics` exposes a limited set of VM and host metrics to the

Reviewer 2: ```suggestion
`downwardMetrics` expose a limited set of VM and host metrics to the
```

Author: Done.

guest. The format is compatible with [vhostmd](https://github.com/vhostmd/vhostmd).

Getting a limited set of host and VM metrics is in some cases required to allow
third-parties diagnosing performance issues on their appliances. One prominent
example is SAP HANA.

> In order to expose the `downwardMetrics` to the VMs, there are two methods supported: `disk` and `virtio-serial port`.

Reviewer 2: ```suggestion
In order to expose `downwardMetrics` to VMs, the two methods `disk` and `virtio-serial port` are supported.
```

Author: Done.

> **Note:** The **DownwardMetrics** feature gate
> [must be enabled](../operations/activating_feature_gates.md#how-to-activate-a-feature-gate)
> to use the metrics. Available starting with KubeVirt v0.42.0.
 
#### Disk

> A volume is created, and it is exposed to the guest as a raw block volume.

Reviewer 2: Add a note that KubeVirt is writing downward metrics periodically into this volume and that the guest can read them?

Author: Addressed. Thanks for the suggestion, missed this detail.

Example:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-fedora
  name: vmi-fedora
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: metrics
    machine:
      type: ""
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - name: containerdisk
    containerDisk:
      image: quay.io/containerdisks/fedora:latest
  - name: metrics
    downwardMetrics: {}
```

#### Virtio-serial port

This method uses a virtio-serial port to expose the metrics data to the VM. 

> It creates a port named `/dev/virtio-ports/org.github.vhostmd.1` inside the VM, from where the metrics can be retrieved.

Reviewer 2: Make it more clear that KubeVirt implements/supports the Virtio Transport protocol of vhostmd over this port?

Author: Done!

The virtio-serial creates a port named `org.github.vhostmd.1` inside the VM. The metrics may be requested and fetched 

> through this port.

Reviewer: what about: 

This method uses a virtio-serial port to expose the metrics data to the VM. 
It creates a port named `/dev/virtio-ports/org.github.vhostmd.1` inside the VM, 
from where the metrics can be retrieved.
See [vhostmd documentation](https://github.com/vhostmd/vhostmd/blob/master/README) under `Virtio Transport` for further information.

Author: I think it's much cleaner with your suggestion. Many thanks! Addressed.

To enable the `downwardMetrics` within a virtio-serial port you need to add a `downwardMetrics` device i.e., 

> the spec `spec.domain.devices.downwardMetrics: {}` has to be added to the VM definition.

Reviewer: what about:

To expose the metrics using a virtio-serial port, a `downwardMetrics` device must be added (i.e., 
`spec.domain.devices.downwardMetrics: {}`).

Author: Done! Thanks for the suggestion :)

Example:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-fedora
  name: vmi-fedora
spec:
  domain:
    devices:
      downwardMetrics: {}
      disks:
      - disk:
          bus: virtio
        name: containerdisk
    machine:
      type: ""
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - name: containerdisk
    containerDisk:
      image: quay.io/containerdisks/fedora:latest
```
#### Accessing Metrics Data

To access the DownwardMetrics shared with a disk or a virtio-serial port, the `vm-dump-metrics` tool can be used:

```xml
$ sudo dnf install -y vm-dump-metrics
$ sudo vm-dump-metrics
<metrics>
  <metric type="string" context="host">
    <name>HostName</name>
    <value>node01</value>
[...]
  <metric type="int64" context="host" unit="s">
    <name>Time</name>
    <value>1619008605</value>
  </metric>
  <metric type="string" context="host">
    <name>VirtualizationVendor</name>
    <value>kubevirt.io</value>
  </metric>
</metrics>
```

> Nevertheless, the application that consumes the metrics, if any, usually will connect by itself to the virtio-serial port

Reviewer 2: ```suggestion
`vm-dump-metrics` is useful as a standalone tool to verify the serial port is working. Applications that consume metrics will usually connect to the virtio-serial port themselves however.
```

Author: Done!

### downwardMetrics

`downwardMetrics` expose a limited set of VM and host metrics to the
guest. The format is compatible with [vhostmd](https://github.com/vhostmd/vhostmd).

Getting a limited set of host and VM metrics is in some cases required to allow
third-parties diagnosing performance issues on their appliances. One prominent
example is SAP HANA.

In order to expose `downwardMetrics` to VMs, the methods `disk` and `virtio-serial port` are supported.

> **Note:** The **DownwardMetrics** feature gate
> [must be enabled](../operations/activating_feature_gates.md#how-to-activate-a-feature-gate)
> to use the metrics. Available starting with KubeVirt v0.42.0.
 
#### Disk

> A volume is created, and it is exposed to the guest as a raw block volume that can be read and KubeVirt
> updates periodically (by default, every 5 seconds).

Reviewer 3: This seems like it should be two sentences, here's a possible breakdown:

```suggestion
A volume is created and it is exposed to the guest as a raw block volume. 
KubeVirt will update it periodically (by default, every 5 seconds).
```

Author: Nice suggestion, like it. Thanks! Addressed.

Example:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-fedora
  name: vmi-fedora
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: metrics
    machine:
      type: ""
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - name: containerdisk
    containerDisk:
      image: quay.io/containerdisks/fedora:latest
  - name: metrics
    downwardMetrics: {}
```

#### Virtio-serial port

This method uses a virtio-serial port to expose the metrics data to the VM.
KubeVirt creates a port named `/dev/virtio-ports/org.github.vhostmd.1` inside the VM, in which the Virtio Transport protocol

> is supported. `donwardMetrics` can be retrieved from this port.

Reviewer 2: typo: `donwardMetrics

Author: Nice catch! Thanks. Addressed.
